### PR TITLE
Python host filter policy PYTHON-761

### PR DESCRIPTION
**Pull request by mambocab**
I need to write more documentation before we merge, but I wanted to get discussion going on API choices.

I chose to model the API here almost exactly off the [Java driver implementation](https://github.com/datastax/java-driver/blob/3.x/driver-core/src/main/java/com/datastax/driver/core/policies/HostFilterPolicy.java). The two relevant choices here:

- the predicate is immutable, and
- host-filtering takes place at policy-population time, not at query-planning time.

Does anyone have strong feelings either way about those? Changing either of these is totally doable, but would add more complexity and make the class harder to test.

### CODE REVIEW

**Comment by mambocab** [comment]
Also -- do we think the one integration test I wrote is enough?

File: cassandra/policies.py [hunk 1]
```diff
@@ -405,12 +410,24 @@ class WhiteListRoundRobinPolicy(RoundRobinPolicy):
...
     def __init__(self, hosts):
         """
         The `hosts` parameter should be a sequence of hosts to permit
         connections to.
         """
+        msg = ('WhiteListRoundRobinPolicy is deprecated. '
+               'It will be removed in 4.0.')
+        warn(msg, DeprecationWarning)
+        # DeprecationWarnings are silent by default so we also log the message
+        log.warning(msg)
```
Comment (aboudreault) [inline cassandra/policies.py]: I am not sure it is our job to ensure people are seeing deprecated warnings. Are we doing that somewhere else in the code base?

File: cassandra/policies.py [hunk 2]
```diff
@@ -441,6 +458,85 @@ def on_add(self, host):
...
+    is truthy. Hosts for which ``predicate(host)`` is falsey will be considered
+    :attr:`.IGNORED`, and will not be used in a query plan.
+    """
+
+    _predicate = None
+
+    def __init__(self, child_policy, predicate):
+        super(HostFilterPolicy, self).__init__()
+        self._child_policy = child_policy
+        self._predicate = predicate
```
Comment (aboudreault) [inline cassandra/policies.py]: we could probably check if the predicate is callable here and raise an error if not.

File: tests/unit/test_policies.py [hunk 3]
```diff
@@ -19,6 +19,10 @@
 
 from itertools import islice, cycle
 from mock import Mock, patch
+try:
+    from thread import LockType  # py2
+except ImportError:
+    from _thread import LockType  # py3
```
Comment (aboudreault) [inline tests/unit/test_policies.py]: six.moves._thread

File: cassandra/policies.py [hunk 4]
```diff
@@ -441,6 +458,85 @@ def on_add(self, host):
...
 
 
+class HostFilterPolicy(LoadBalancingPolicy):
+    """
+    A :class:`.LoadBalancingPolicy` subclass that takes an instantiated
+    ``LoadBalancingPolicy`` as a child policy, and a single-argument predicate.
+    This policy defers to the child policy for hosts where ``predicate(host)``
+    is truthy. Hosts for which ``predicate(host)`` is falsey will be considered
+    :attr:`.IGNORED`, and will not be used in a query plan.
+    """
```
Comment (aboudreault) [inline cassandra/policies.py]: I think it might be worth adding an example here. ie. the old WhiteListRoundRobinPolicy done with this new one.

**Comment by aboudreault** [comment]
LGTM Jim! Minor comments added.

File: tests/unit/test_policies.py (hunk 3 above)
Comment (mambocab) [inline tests/unit/test_policies.py]: Hah -- I swear I looked for this. Thanks

File: cassandra/policies.py (hunk 1 above)
Comment (mambocab) [inline cassandra/policies.py]: We are -- for instance:

https://github.com/datastax/python-driver/commit/9421d36de79ede7fdb20873fe5b117ee2d05b2e9#diff-9e9c9d4aa3f0469896e27bf747ef9e82R126

IMO the default is crappy enough that we should work around it.

File: cassandra/policies.py (hunk 2 above)
Comment (mambocab) [inline cassandra/policies.py]: I might slightly prefer `except TypeError` when calling `self._predicate` and raising an error with a more useful error message. It's slightly more idiomatic, with the whole whole "better to ask forgiveness than permission" thing. Do you think one has technical merits over the other?

**Comment by mambocab** [comment]
Oh, yeah, Alan -- did you have feelings either way about the implementation choices I mentioned in my request?

The one problem I can think of with the Java driver's approach is that `distance` calculates `predicate(host) when it's called, while `make_query_plan` uses data calculated during `on_{up,down,add,remove}`. It's a little weird that the results can be different if the conditions that trip the predicate change.

File: cassandra/policies.py (hunk 4 above)
Comment (aboudreault) [inline cassandra/policies.py]: Also, I think the parameters should be documented in the __init__ function, so that autodoc generate that properly.

**Comment by aboudreault** [comment]
Jim, I think what we have is correct. The hosts list should only change if we call populate (or on_up etc.), not at the query plan time. I don't see where Java does this, can you point me to the right place?

**Comment by mambocab** [comment]
> The hosts list should only change if we call populate (or on_up etc.), not at the query plan time. I don't see where Java does this, can you point me to the right place?

You're correct -- the host list only changes in those situations, and `make_query_plan` defers to the child policy, so it uses results calculated during those host list changes. However, `.distance` doesn't use this result; it calculates `predicate(host)` when called.

So, problems can happen if `host` changes in some way (other than being added/removed or going up/down), or if `predicate` depends on some other kind of state that changes. You could run into a situation where, for instance, a `predicate(host)` is `IGNORED`, but is returned in `make_query_plan`. Is this ok? (I also want to mention this to the Java driver devs to see what they think.)

**Comment by mambocab** [comment]
The latest commit implements this:

https://github.com/datastax/python-driver/pull/778#discussion_r120480828

in what might be a pretty gross way -- could you let me know your thoughts, Alan?

The code could be simpler if we didn't care about providing a public, read-only `.predicate` attr, but I do want to provide that so clients can check the predicate themselves without accessing private attribute.

I believe I've addressed your other comments. I'd like to wait on a decision for  [JAVA-1489](https://datastax-oss.atlassian.net/browse/JAVA-1489) before finalizing our decision on when methods call `.predicate`.

File: cassandra/policies.py [hunk 5]
```diff
@@ -441,6 +460,133 @@ def on_add(self, host):
...
+
+        @wraps(predicate)
+        def predicate_with_better_error_handling(*args, **kwargs):
+            try:
+                return predicate(*args, **kwargs)
+            except TypeError as te:
+                msg = (
+                    'caught the following error while calling ' +
+                    self.__class__.__name__ + " object's predicate method: " +
+                    repr(te.message)
```
Comment (aboudreault) [inline cassandra/policies.py]: I don't think te.message is available for Python3

Comment (mambocab) [inline cassandra/policies.py]: grrrrr

**Comment by aboudreault** [comment]
@mambocab Is there anything that force us to use your public accessor for the predicate? ie:
```
self.predicate(hola)
```

IMO, I think it would be more clean to create a private function that basically call the predicate and do your error handling then use it internally. ie. self._call_predicate(hola). Now, I feel we are scarifying our code for a feature we aren't sure that will be used (the predicate property). What do you think Jim? Feel free to take the final decision on that.

**Comment by mambocab** [comment]
> Now, I feel we are scarifying our code for a feature we aren't sure that will be used (the predicate property). What do you think Jim? Feel free to take the final decision on that.

I kinda like continuing to expose `.predicate` for debugging. I've removed that wild error-handling wrapping code, I'm not sure why I thought it was necessary. Errors thrown by the predicate will show that they were thrown while calling `predicate`.

**Comment by mambocab** [comment]
@aboudreault Are you happy with that solution? I think the existing error reporting is actually ok.

**Comment by aboudreault** [comment]
LGTM!

**Comment by mambocab** [comment]
Thanks for the review. I've cleaned up an unused import and added a test for the deprecation warning, so now I'll squash and merge.

**Comment by mambocab** [comment]
Rebasing https://github.com/datastax/python-driver/pull/778/commits/c220991940dc3c6a19c003847d81cf51ae003708 down to a pre-HFP whitespace commit and a commit for HFP.

**Pull request merged**